compiler/rustc_lint/src/lifetime_syntax.rs

```diff
@@ -527,3 +653,115 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeInfoCollector<'a, 'tcx> {
         self.referenced_type_span = old_referenced_type_span;
     }
 }
+
+declare_lint! {
+    /// The `hidden_lifetimes_in_type_paths` lint detects the use of
+    /// hidden lifetime parameters in types not part of a function's
+    /// arguments or return values.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(hidden_lifetimes_in_type_paths)]
+    ///
+    /// struct ContainsLifetime<'a>(&'a i32);
+    ///
+    /// static FOO: ContainsLifetime = ContainsLifetime(&42);
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Hidden lifetime parameters can make it difficult to see at a
+    /// glance that borrowing is occurring.
+    ///
+    /// This lint ensures that lifetime parameters are always
+    /// explicitly stated, even if it is the `'_` [placeholder
+    /// lifetime].
+    ///
+    /// [placeholder lifetime]: https://doc.rust-lang.org/reference/lifetime-elision.html#lifetime-elision-in-functions
+    pub HIDDEN_LIFETIMES_IN_TYPE_PATHS,
+    Allow,
+    "hidden lifetime parameters in types outside function signatures are discouraged"
```

Comment:
Suggestion [WORDING-STRENGTH 2/2]: same here, maybe lessen the "strength" conveyed in *this* initial PR adding this lint as allow-by-default, and defer the lint level bump + wording strengthening (to "deprecate" and "discourage") in the follow-up upgrade-to-warn-by-default PR. I.e. maybe

> hidden lifetime parameters in types outside function signatures can be confusing

But I don't feel *too* strongly about this.